src/rt/rwlock.rs

```diff
@@ -0,0 +1,269 @@
+use crate::rt::{
+    object::{self, Object},
+    thread, Access, Synchronize,
+};
+
+use std::collections::HashSet;
+use std::sync::atomic::Ordering::{Acquire, Release};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(crate) struct RwLock {
+    obj: Object,
+}
+
+#[derive(Debug, PartialEq)]
+enum RwLockType {
+    Read(HashSet<thread::Id>),
+    Write(thread::Id),
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    /// If the rwlock should establish sequential consistency.
+    seq_cst: bool,
+
+    /// `Some` when the rwlock is in the locked state.
+    /// The `thread::Id` references the thread that currently holds the rwlock.
+    lock: Option<RwLockType>,
+
+    /// Tracks write access to the rwlock.
+    last_access: Option<Access>,
+
+    /// Causality transfers between threads
+    synchronize: Synchronize,
+}
+
+impl RwLock {
+    /// Common RwLock function
+    pub(crate) fn new(seq_cst: bool) -> RwLock {
```

Comment:
As far as I can tell, the `rt::RwLock` type is only ever constructed in `sync::RwLock::new`, and it always passes `true` here. Will we ever construct a `rt::RwLock` with `seq_cst=false`? If not, should we just remove the bool and always establish sequential consistency?